File: src/apps/mobile/src/api/coach.ts
/** Coach endpoints. Types mirror apps/api/app/schemas/coach.py. */
import { api } from './client';

export type CoachCard = {
  kind: 'demand' | 'motif' | 'stale' | 'questions';
  title: string;
  action: string;
  focus: boolean;
  impact_rupees: number;
  body?: string | null;
  fromPrice?: string | null;
  toPrice?: string | null;
  product?: string | null;
  fix?: string | null;
  meta: Record<string, unknown>;
};

export type CoachCardsResponse = {
  cards: CoachCard[];
  generated_at: string;
};

export type CoachDigest = {
  date: string;
  live_listings: number;
  views_total: number;
  orders_today: number;
  revenue_today: number;
  message_en: string;
  message_hi: string;
};

export type MockupTemplate = {
  id: string;
  label: string;
  base_price: number;
};

export type MockupResult = {
  media_id: string;
  url: string;
  template: string;
  label: string;
  current_price: number;
  potential_price: number;
  multiplier: number;
  uplift: number;
};

export const coachApi = {
  /**
   * persist=false on a plain screen-open pull: writing a nudge stamps
   * last_coached_at on the listing it came from, which is what enforces the
   * 14-day quiet period server-side (services/coach.py) -- reopening the
   * Learn tab three times in a morning must not burn through that window.
   */
  getCards: (persist = false) =>
    api.get<CoachCardsResponse>(`coach/cards?persist=${persist}`),

  getDigest: () => api.get<CoachDigest>('coach/digest'),

  getTemplates: () => api.get<MockupTemplate[]>('coach/templates'),

  generateMockup: (template: string, mediaId: string) =>
    api.post<MockupResult>(`coach/mockup/${template}?media_id=${mediaId}`),

  actOnNudge: (nudgeId: string, actedOn = true) =>
    api.patch<{ ok: boolean }>(`coach/nudges/${nudgeId}`, { acted_on: actedOn }),
};

File: src/apps/api/app/schemas/coach.py
"""Coach schemas -- phase 10."""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel


class CoachCardOut(BaseModel):
    """
    One card, shaped to match CoachCard in
    apps/mobile/src/data/sampleContent.ts exactly -- the Learn screen already
    renders this shape against sample data; this is what makes swapping the
    sample constant for a real fetch a one-line change, not a rewrite.
    """

    kind: Literal["demand", "motif", "stale", "questions"]
    title: str
    action: str
    focus: bool = False
    impact_rupees: int = 0
    # Present only on their matching kind -- optional here, required in the
    # union sampleContent.ts defines, because the API sends one flat shape
    # and lets absent fields mean "not this kind" rather than modelling a
    # discriminated union twice.
    body: str | None = None
    fromPrice: str | None = None
    toPrice: str | None = None
    product: str | None = None
    fix: str | None = None
    meta: dict[str, Any] = {}


class CoachCardsOut(BaseModel):
    cards: list[CoachCardOut]
    generated_at: str


class DigestOut(BaseModel):
    date: str
    live_listings: int
    views_total: int
    orders_today: int
    revenue_today: float
    message_en: str
    message_hi: str


class TemplateOut(BaseModel):
    id: str
    label: str
    base_price: int


class MockupOut(BaseModel):
    media_id: str
    url: str
    template: str
    label: str
    current_price: int
    potential_price: int
    multiplier: float
    uplift: int


class NudgeActionIn(BaseModel):
    acted_on: bool = True
